enhancements/hypershift/priority-class/priority-class-dynamic-adjustment.md

```diff
@@ -0,0 +1,251 @@
+
+creation-date: 2022-09-23
+last-updated: 2023-04-10
+---
+# Priority Classes
+
+## Summary
+
+This proposal details the plan for enabling the HyperShift Operator to adjust the priority class of master control plane components to be higher if the customer is using more worker nodes in their cluster. 
+
+## Motivation
+The goal of this effort is to add support for the HyperShift Operator to ensure minimal disruption for customer control plane workloads that generate higher revenue. Priority classes on the control plane workloads can be dynamically adjusted based on the number of worker nodes associated to the HC.
+This helps ensure that control plane pods of higher priority clusters don't get evicted/rescheduled before the pods of a lower priority cluster. Dynamically adjusting priority classes is something IBM Cloud has been doing to maintain and operate their Kubernetes/OpenShift service at scale.
+
+### User Stories
+- As a Service Provider, I want to ensure the reliability of our service offering for our paying customers.
+- As a Service Provider, I want to have priority classes dynamically adjusted based on the number of worker nodes being used by a cluster.
+- As a Service Provider, I want the ability to override priority classes if needed for debugging or any other operational tasks.
+- As a Service Consumer, I want to have confidence in the reliability of the service so my applications are not disrupted.
+- As IBM, I want to ensure priority class adjustments are supported for HyperShift-based clusters to maintain our operation of Kubernetes/OpenShift offerings at scale.
+
+### Goals
+- Define how new priority classes are created on the management cluster.
+  - Option 1: Default sets to be rendered by HyperShift CLI tool at install time.
+  - Option 2: User-provided by management cluster admins.
+- Expose priority class name being used for a given HC to integrate with external systems.
+  - e.g. IBM adjusts resource requests and limits based on the priority class being set.
+- Define how HO will determine "priority" for a given HC.
+  - e.g. IBM determines this based on customer investment (cluster size/worker node count).
+- Ensure dynamic priority class allocation is available by default
+  - Clients should be able to opt out of employing the features offered through this enhancement.  
+- Define sensible priority class recommendations.
+  - ETCD should always have a higher priority class to account for unwarranted evictions.
+
+### Non-Goals
+- HO will not adjust any pod resource request and limits.
+- Priority class adjustments to a lower priority value will not be supported.
+
+## Proposal
+HyperShift currently creates a static set of default priority classes. We propose to add support for dynamically adjusting priority classes for the underlying control plane workloads. This will help to ensure the HCs with a higher level of usage/investment will be of more importance when scheduling workloads and mitigating preemptions. 
```

Comment:
Does this imply dynamic creation of priority class to accomidate arbitrarily large hosted clusters?